src/topology/inseparable.lean

```diff
@@ -178,12 +179,205 @@ def specialization_preorder : preorder X :=
+  stable_under_specialization sᶜ ↔ stable_under_generalization s :=
+is_lower_set_compl
+
+lemma stable_under_generalization_compl_iff {s : set X} :
+  stable_under_generalization sᶜ ↔ stable_under_specialization s :=
+is_upper_set_compl
+
+lemma stable_under_specialization_sUnion (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_specialization s) : stable_under_specialization (⋃₀ S) :=
+is_lower_set_sUnion H
+
+lemma stable_under_specialization_sInter (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_specialization s) : stable_under_specialization (⋂₀ S) :=
+is_lower_set_sInter H
+
+lemma stable_under_generalization_sUnion (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_generalization s) : stable_under_generalization (⋃₀ S) :=
+is_upper_set_sUnion H
+
+lemma stable_under_generalization_sInter (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_generalization s) : stable_under_generalization (⋂₀ S) :=
+is_upper_set_sInter H
+
+lemma stable_under_specialization_Union {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_specialization (S i)) : stable_under_specialization (⋃ i, S i) :=
+is_lower_set_Union H
+
+lemma stable_under_specialization_Inter {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_specialization (S i)) : stable_under_specialization (⋂ i, S i) :=
+is_lower_set_Inter H
+
+lemma stable_under_generalization_Union {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋃ i, S i) :=
+is_upper_set_Union H
+
+lemma stable_under_generalization_Inter {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋂ i, S i) :=
+is_upper_set_Inter H
+
+lemma stable_under_specialization_iff_eq_Union {s : set X} :
+  stable_under_specialization s ↔ s = ⋃ x ∈ s, closure {x} :=
+by simpa only [closure_singleton_eq_Iic] using is_lower_set_iff_eq_Inter_Iic
+
+alias stable_under_specialization_iff_eq_Union ↔ stable_under_specialization.eq_Union _
+
+/-- A set is stable under specialization iff it is a union of closed sets. -/
+lemma stable_under_specialization_iff_exists_sUnion_eq {s : set X} :
+  stable_under_specialization s ↔ ∃ (S : set (set X)), (∀ s ∈ S, is_closed s) ∧ ⋃₀ S = s :=
+begin
+  refine ⟨λ H, ⟨(λ x : X, closure {x}) '' s, _, _⟩, λ ⟨S, hS, e⟩, e ▸
+    stable_under_specialization_sUnion S (λ x hx, (hS x hx).stable_under_specialization)⟩,
+  { rintros _ ⟨_, _, rfl⟩, exact is_closed_closure },
+  { conv_rhs { rw stable_under_specialization_iff_eq_Union.mp H }, simp }
+end
+
+/-- A set is stable under generalization iff it is an intersection of open sets. -/
+lemma stable_under_generalization_iff_exists_sInter_eq {s : set X} :
+  stable_under_generalization s ↔ ∃ (S : set (set X)), (∀ s ∈ S, is_open s) ∧ ⋂₀ S = s :=
+begin
+  refine ⟨_, λ ⟨S, hS, e⟩, e ▸
+    stable_under_generalization_sInter S (λ x hx, (hS x hx).stable_under_generalization)⟩,
+  rw [← stable_under_specialization_compl_iff,
+    stable_under_specialization_iff_exists_sUnion_eq],
```

Comment:
```suggestion
  rw [← stable_under_specialization_compl_iff, stable_under_specialization_iff_exists_sUnion_eq],
```

Reply:
I don't think this is right?

Reply:
? I mean the whole `rw` fits in one line

Reply:
Oh the display made me think that you're removing the indent.